agent/src/storage/mod.rs

```diff
@@ -1,80 +1,348 @@
+    pub async fn data_exists_by_key(&self, fq_name: &str) -> ServalResult<bool> {
+        let key = Manifest::make_manifest_key(fq_name);
+
+        // If we make a successful local check, we return only if we found it.
+        // We're going to fall back to bucket storage if we have it.
+        if let Some(local) = &self.local {
+            if let Ok(v) = local.data_exists_by_key(&key).await {
+                if v {
+                    return Ok(v);
                 }
             }
-            // If we get here we have utterly failed and cannot continue, but crashing might not be right.
-            Err(ServalError::StorageError(
-                "We were unable to find any peers with the storage role on this mesh.".to_string(),
-            ))
         }
+
+        if let Some(bucket) = &self.bucket {
+            if let Ok(v) = bucket.data_exists_by_key(&key).await {
+                return Ok(v);
+            }
+        }
+
+        // This is not exposed in the API yet so we cannot proxy, so we admit defeat.
+        Ok(false)
     }
-}
 
-/// This trait expresses the duties of a storage implementation that meets the requirments of
-/// a Wasm job runner node.
-#[async_trait]
-pub trait RunnerStorage {
     /// Fetch a manifest by its fully-qualified name.
-    async fn manifest(&self, fq_name: &str) -> Result<Manifest, ServalError>;
+    pub async fn manifest(&self, fq_name: &str) -> ServalResult<Manifest> {
+        if !self.has_storage() {
+            let proxy = make_proxy_client().await?;
+            return proxy.get_manifest(fq_name).await;
+        }
 
-    /// Fetch the bytes of the named executable so we can run it.
-    async fn executable_as_bytes(&self, name: &str, version: &str) -> Result<Vec<u8>, ServalError>;
+        if let Some(local) = &self.local {
+            if let Ok(v) = local.manifest(fq_name).await {
+                return Ok(v);
+            }
+        }
 
-    // The following three functions are speculative implementations of things that
-    // that runner nodes do not need today, but that were easy to implement just in case.
+        if let Some(bucket) = &self.bucket {
+            if let Ok(v) = bucket.manifest(fq_name).await {
+                return Ok(v);
+            }
+        }
+
+        Err(ServalError::ManifestNotFound(fq_name.to_string()))
+    }
 
     /// Store a Wasm manifest. Returns the integrity checksum.
-    async fn store_manifest(&self, manifest: &Manifest) -> Result<Integrity, ServalError>;
+    pub async fn store_manifest(&self, manifest: &Manifest) -> ServalResult<Integrity> {
+        if !self.has_storage() {
+            let proxy = make_proxy_client().await?;
+            return proxy.store_manifest(manifest).await;
+        }
+
+        let local_result = if let Some(local) = &self.local {
+            Some(local.store_manifest(manifest).await)
+        } else {
+            None
+        };
+
+        let bucket_result = if let Some(bucket) = &self.bucket {
```

Comment:
I may be trying to be too fancy, but can this be done as a `map` to avoid the if/else? As in: I am still learning Rust, literally can we even do this? If not, why not? "It's not idiomatic Rust" is a very useful response.

(Computers are fun! This is intended as a friendly conversation, not a nitpick.)

````rs
let bucket_result = &self.bucket.map(|bucket| bucket.store_manifest(manifest).await)```